### ISSUE TITLE
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

### ISSUE THREAD

**Issue by chenziyi** [dylkot/cNMF#73]
Dear Dylan,

I run cnmf with the adata transformed from Seurat with function 'Convert' 

Convert(Seurat_data.combined.rename.h5seurat", assay='SCT', dest = "h5ad")

After that, I start to preprossess it in python with the following codes:
adata=sc.read_h5ad( "Seurat_data.combined.rename.h5ad")
del adata.raw
adata.X = adata.X.astype(np.float64)
os.system('/data/software/anaconda3/envs/scrnaseq/bin/cnmf prepare --output-dir /data/project/cNMF --name cNMF -c /data/project/cNMF/Seurat_data.combined.rename.h5ad -k 5 6 7 8 9 10 --n-iter 200 --total-workers 4 --seed 14 --numgenes 2000 --beta-loss frobenius')
 and the following errors reported.

/data/software/anaconda3/envs/scrnaseq/lib/python3.11/site-packages/scanpy/preprocessing/_simple.py:843: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
Traceback (most recent call last):
  File "/data/software/anaconda3/envs/scrnaseq/bin/cnmf", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/data/software/anaconda3/envs/scrnaseq/lib/python3.11/site-packages/cnmf/cnmf.py", line 1027, in main
    cnmf_obj.prepare(args.counts, components=args.components, n_iter=args.n_iter, densify=args.densify,
  File "/data/software/anaconda3/envs/scrnaseq/lib/python3.11/site-packages/cnmf/cnmf.py", line 359, in prepare
    norm_counts = self.get_norm_counts(input_counts, tpm, num_highvar_genes=num_highvar_genes,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/software/anaconda3/envs/scrnaseq/lib/python3.11/site-packages/cnmf/cnmf.py", line 458, in get_norm_counts
    examples = norm_counts.obs.index[zerocells]
               ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/data/software/anaconda3/envs/scrnaseq/lib/python3.11/site-packages/pandas/core/indexes/base.py", line 5196, in __getitem__
    result = getitem(key)
             ^^^^^^^^^^^^
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed


Could you please help me for this issue?

**Comment by twoneu** [comment]
Hi @dylkot, I'm also experiencing this error - do you have any advice?

**Comment by wchenqi** [comment]
do you figure it out? I have the same problem. Would you please share the solution~~ Thank you so much!

**Comment by jamesnemesh** [comment]
I've seen an error that looks like this before - this may be due to some cells having very small expression values.   

One thing about cNMF that's important to remember is that the NMF part runs on highly variable genes.  If the program picks out 2000 genes and a cell has 0 expression across those 2000 genes, then it will eventually fail.  The function it's dying in is supposed to detect that.  

You should pre-filter your data to remove genes with very low expression and cells with very small numbers of UMIs.  For example, you could use the following function: 

```
def filter_data(adata, min_genes=200, min_counts=200, min_cells_per_gene=10):
    """Filter the DGE matrix to remove genes/cells with low expression

    Parameters
    ------------------
    adata: AnnData
        An AnnData object with gene expression measurements
    min_cells_per_gene: integer
        The minimum number of cells that must express a gene
    min_genes: integer
        The minimim number of genees that a cell must express with at least one UMI
    min_counts: integer
        The minimum number of total UMIs the cell must have (library size) to be included in analysis
    """

    print("Starting cells before filtering %d" % adata.n_obs)
    sc.pp.filter_cells(adata, min_genes=min_genes)  # filter cells with fewer than [x] genes
    sc.pp.filter_cells(adata,
                       min_counts=min_counts)  # This is a weaker threshold than above. It is just to population the n_counts column in adata
    print("Cells after filtering %d" % adata.n_obs)
    print("Starting genes before filtering low expression [gene seen in less than %d cells] %d" % (
        min_cells_per_gene, adata.n_vars))
    sc.pp.filter_genes(adata, min_cells=min_cells_per_gene)  # filter genes detected in fewer than [x] cells
    print("Genes after filtering low expression [gene seen in less than %d cells] %d" % (
        min_cells_per_gene, adata.n_vars))

    # Filter MT genes
    print("Starting genes before MT filtering %d" % adata.n_vars)
    mtGenes = adata.var_names.str.startswith('MT-')
    adata = adata[:, ~mtGenes]
    print("Genes after MT filtering %d" % adata.n_vars)
    return (adata)
```

I've hacked up cNMF quite a bit on my private repository, I could have sworn that this was specific to my code and I had made a mistake (and I just fixed this in my own copy yesterday), but maybe it's always been there.

There's a silly mistake in the line that's breaking, where the zerocells is actually of size (N,) where as the pandas index needs a vector of length (N).

```
examples = norm_counts.obs.index[zerocells]
```

The code in that block should be: 
```
zero_cells = norm_counts.X.sum(axis=1) == 0
        if zero_cells.sum() > 0:
            index_zero_cells = np.ravel(zero_cells)
            examples = norm_counts.obs.index[index_zero_cells]
            print(
                "Warning: %d cells have zero counts of overdispersed genes. E.g. %s"
                % (zero_cells.sum(), examples[0]),
                flush=True,
            )
            print(
                "Consensus step will not run when this is the case.  Filter genes/cells more to fix this error.  Quitting!",
                flush=True,
            )
            sys.exit(1)
        return norm_counts
```

**Comment by wchenqi** [comment]
Thank you!!! I'll go try re-filter my data.
| |
cenqi951108
|
|
***@***.***
|
---- Replied Message ----
| From | ***@***.***> |
| Date | 12/7/2023 03:44 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [dylkot/cNMF] IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed (Issue #73) |

I've seen an error that looks like this before - this may be due to some cells having very small expression values.

One thing about cNMF that's important to remember is that the NMF part runs on highly variable genes. If the program picks out 2000 genes and a cell has 0 expression across those 2000 genes, then it will eventually fail. The function it's dying in is supposed to detect that.

You should pre-filter your data to remove genes with very low expression and cells with very small numbers of UMIs. For example, you could use the following function:

def filter_data(adata, min_genes=200, min_counts=200, min_cells_per_gene=10):
    """Filter the DGE matrix to remove genes/cells with low expression

    Parameters
    ------------------
    adata: AnnData
        An AnnData object with gene expression measurements
    min_cells_per_gene: integer
        The minimum number of cells that must express a gene
    min_genes: integer
        The minimim number of genees that a cell must express with at least one UMI
    min_counts: integer
        The minimum number of total UMIs the cell must have (library size) to be included in analysis
    """

    print("Starting cells before filtering %d" % adata.n_obs)
    sc.pp.filter_cells(adata, min_genes=min_genes)  # filter cells with fewer than [x] genes
    sc.pp.filter_cells(adata,
                       min_counts=min_counts)  # This is a weaker threshold than above. It is just to population the n_counts column in adata
    print("Cells after filtering %d" % adata.n_obs)
    print("Starting genes before filtering low expression [gene seen in less than %d cells] %d" % (
        min_cells_per_gene, adata.n_vars))
    sc.pp.filter_genes(adata, min_cells=min_cells_per_gene)  # filter genes detected in fewer than [x] cells
    print("Genes after filtering low expression [gene seen in less than %d cells] %d" % (
        min_cells_per_gene, adata.n_vars))

    # Filter MT genes
    print("Starting genes before MT filtering %d" % adata.n_vars)
    mtGenes = adata.var_names.str.startswith('MT-')
    adata = adata[:, ~mtGenes]
    print("Genes after MT filtering %d" % adata.n_vars)
    return (adata)


I've hacked up cNMF quite a bit on my private repository, I could have sworn that this was specific to my code and I had made a mistake (and I just fixed this in my own copy yesterday), but maybe it's always been there.

There's a silly mistake in the line that's breaking, where the zerocells is actually of size (N,) where as the pandas index needs a vector of length (N).

examples = norm_counts.obs.index[zerocells]


The code in that block should be:

zero_cells = norm_counts.X.sum(axis=1) == 0
        if zero_cells.sum() > 0:
            index_zero_cells = np.ravel(zero_cells)
            examples = norm_counts.obs.index[index_zero_cells]
            print(
                "Warning: %d cells have zero counts of overdispersed genes. E.g. %s"
                % (zero_cells.sum(), examples[0]),
                flush=True,
            )
            print(
                "Consensus step will not run when this is the case.  Filter genes/cells more to fix this error.  Quitting!",
                flush=True,
            )
            sys.exit(1)
        return norm_counts

**Comment by dylkot** [comment]
Hey guys,

Sorry about this issue. I will correct the code shortly to address this issue.

Thanks!

**Comment by dylkot** [comment]
Hey all this issue should now be fixed on the master branch and on pypi in 1.4.1. Let me know if there are any issues. Thanks @jamesnemesh!
